lld/test/COFF/stub.test

```diff
@@ -0,0 +1,33 @@
+# RUN: yaml2obj %p/Inputs/ret42.yaml -o %t.obj
+
+# RUN: lld-link /out:%t.exe /entry:main /stub:%p/Inputs/stub512mz %t.obj
+# RUN: llvm-readobj --file-headers %t.exe | FileCheck -check-prefix=CHECK1 %s
+
+CHECK1: Magic: MZ
+CHECK1: UsedBytesInTheLastPage: 144
+CHECK1: FileSizeInPages: 3
+CHECK1: NumberOfRelocationItems: 0
+CHECK1: HeaderSizeInParagraphs: 4
+CHECK1: MinimumExtraParagraphs: 0
+CHECK1: MaximumExtraParagraphs: 65535
+CHECK1: InitialRelativeSS: 0
+CHECK1: InitialSP: 184
+CHECK1: Checksum: 0
+CHECK1: InitialIP: 0
+CHECK1: InitialRelativeCS: 0
+CHECK1: AddressOfRelocationTable: 64
+CHECK1: OverlayNumber: 0
+CHECK1: OEMid: 0
+CHECK1: OEMinfo: 0
+CHECK1: AddressOfNewExeHeader: 64
+
+## Invalid DOS signature (must be `MZ`)
+# RUN: not lld-link /out:%t.exe /entry:main /stub:%p/Inputs/stub512zz %t.obj 2>&1 | FileCheck -check-prefix=CHECK-INVALID %s
+
+## Unaligned stub (must be aligned to 8)
+# RUN: not lld-link /out:%t.exe /entry:main /stub:%p/Inputs/stub516mz %t.obj 2>&1 | FileCheck -check-prefix=CHECK-INVALID %s
+
+## Too-small stub (must be at least 64 bytes long)
```

Comment:
Wouldn't this case also fail due to not being aligned to 8 bytes? So without seeing the implementation (or looking at an error message telling the exact reason), we don't strictly know if we're hitting the intended error, or erroring out due to some other reason.

Also, the numbers in the file names, 511/512/516 are quite puzzling how they relate to the actual file sizes (63/64/68 bytes). Can you enlighten me, or rename them to include the actual file size in the name instead :-)

Reply:
> Wouldn't this case also fail due to not being aligned to 8 bytes? So without seeing the implementation (or looking at an error message telling the exact reason), we don't strictly know if we're hitting the intended error, or erroring out due to some other reason.

Too-small stub (63 bytes) is not aligned to 8 bytes, yes.

For the latter, I was thinking the similar. Maybe it's better to put an exact cause for each malformed patterns. For now I followed the way how LINK.exe puts an error (they only put "invalid format" in any of the case). Should I do that?

Reply:
Would be something like:

```cpp
  if (bufferSize < 0x40)
    Err(ctx) << "/stub: stub must be greater than or equal to 64 bytes: " << path;
  if (bufferSize % 8 != 0)
    Err(ctx) << "/stub: stub must be aligned to 8 bytes: " << path;
  if ((bufferStart[0] != 'M' || bufferStart[1] != 'Z'))
    Err(ctx) << "/stub: invalid DOS signature: " << path;
```

Reply:
It could probably be valuable to add more value in the error messages, even if link.exe doesn't; we don't need to match link.exe bit-by-bit exact in such details anyway. (For some larger cases, we print matching LNK123 error numbers, but most of other warnings/errors are different anyway.)

If we can grep for the right error message, this is probably fine, but otherwise it'd be good to make the file e.g. 56 bytes, so that it does pass the alignment test, so we trigger specifically the error we want to test, regardless of in which order the checks are made.

Reply:
That sounds reasonable to me, thanks!